### PR TITLE
Issue/446 websocket trouble

### PR DESCRIPTION
**Pull request by marto83**
I've made MsWebsocketConnection and MsWebsocketTransport public and added a new options class that can be passed to them which allows users to set the websocket send and receive limits. 

Closes: #446 

Unfortunately I found a bug in the way we handle Presence. It can't have been around for some time because a week ago all tests passed but the 250 messages presence test no longer passes. Need to check if something has changed with presence.

### CODE REVIEW

File: src/IO.Ably.Shared/Transport/MsWebSocketConnection.cs [hunk 1]
```diff
@@ -2,60 +2,134 @@
...
+        public MsWebSocketConnection(Uri uri, MsWebSocketOptions socketOptions)
         {
-            Logger = logger;
             _uri = uri;
             ClientWebSocket = new ClientWebSocket();
+
+            if (socketOptions.SendBufferInBytes.HasValue || socketOptions.ReceiveBufferInBytes.HasValue)
+            {
+                var receiveBuffer = socketOptions.ReceiveBufferInBytes ?? (8 * 1024);
+                var sendBuffer = socketOptions.SendBufferInBytes ?? (8 * 1024);
```
Comment (sacOO7) [inline src/IO.Ably.Shared/Transport/MsWebSocketConnection.cs:97]: Can we store the constant value somewhere else? maybe in defaults or just refer it from native defaults from the WebSocket lib

File: src/IO.Ably.Shared/Transport/MsWebSocketConnection.cs [hunk 2]
```diff
@@ -2,60 +2,134 @@
...
+        private readonly Channel<MessageToSend> _sendChannel = Channel.CreateUnbounded<MessageToSend>(
+            new UnboundedChannelOptions()
+            {
+                SingleReader = true,
+                SingleWriter = true,
+            });
 
         internal ClientWebSocket ClientWebSocket { get; set; }
 
         private CancellationTokenSource _tokenSource = new CancellationTokenSource();
```
Comment (sacOO7) [inline src/IO.Ably.Shared/Transport/MsWebSocketConnection.cs:82]: Aren't we using the cancel method on the source from the controller/tokenprovider class?
- I am not sure if it's in the code but not in the PR or if it's being called by respective token consumers.
- Or we don't need it in the first place, this is just a hunch though :p

Comment (sacOO7) [reply src/IO.Ably.Shared/Transport/MsWebSocketConnection.cs:82]: Ok, we already are doing it in a dispose :+1:  Just ignore :p

Comment (sacOO7) [reply src/IO.Ably.Shared/Transport/MsWebSocketConnection.cs:82]: Maybe you can look for places where we might accidentally miss it

**Comment by QuintinWillison** [comment]
@sacOO7 Can you please evolve this:

1. Check that we're only exposing the minimum required to provide these override points. I'm a little concerned this might grow the public API for the client library in areas it does not need. Focus on the original issue for scope analysis.
2. Add documentation to the readme describing how to use the new interfaces / overrides, and why a customer might need to use them.

Thanks.

**Comment by sacOO7** [comment]
Also, @marto83 would you like to look into public interfaces/documentation or should I take it up?

**Comment by marto83** [comment]
@sacOO7 I think for this one it will be easier for me to do it. I'll add a section in readme about sending large messages at a higher rate. I've documented all the classes anyway. We can then merge it.

**Review by QuintinWillison — APPROVED** [review]

**Pull request merged**